internal/contour/handler.go

```diff
@@ -267,3 +269,28 @@ func (e *EventHandler) setStatus(statuses map[dag.Meta]dag.Status) {
 		}
 	}
 }
+
+// WaitForCacheProcess waits for the Cache to be processed by checking the
+// value of "outstanding"
+// Return true if processed fully or false if timed out waiting
+func (e *EventHandler) WaitForCacheProcess(timeout time.Duration) bool {
+	ch := make(chan bool, 1)
+
+	go func() {
+		for e.outstanding > 0 {
```

Comment:
I think this is a data race

Reply:
Any thoughts on how we can check for when the cache queue is done? I have a cluster with 8k services, so those all load up and cause the dag to rebuild a bunch (1600+ times). 

This all doesn't take long, so maybe this check isn't needed but wanted to make sure that the dag that got served up to any Envoy had all the objects already processed.

Reply:
I'm not sure polling outstanding is the right solution, even if we could solve the data race, as in a cluster with a stream of updates e.outstanding may never fall to zero, or it may fall to zero temporarily then rise.

Maybe its worth stepping back and asking the question about the operation's we want to occur before accepting gRPC requests. To my mind they are

1. Receive an initial list of updates from the api server, this is whatever client-go says it is by the "wait for cached" operation on the informer.
2. Initial at least one DAG rebuild cycle
3. Wait for that DAG cycle to be propagated through to the grpc caches.

For 1. I think this is well in hand with you change.
For 2. we have the "eh.UpdateNow" method
For 3. perhaps we need something similar to the eh.Sequence counter that is only tickled when cache updates happen. (eh.Sequence is tickled every time OnAdd/Update/Delete is called, which may not directly relate to cache updates)